pkg/controller/assignmetadata/assignmetadata_controller.go

```diff
@@ -139,38 +171,107 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	deleted = deleted || !assignMetadata.GetDeletionTimestamp().IsZero()
 	tracker := r.tracker.For(gvkAssignMetadata)
 
+	mID, err := types.MakeID(assignMetadata)
+	if err != nil {
+		tracker.TryCancelExpect(assignMetadata)
+		log.Error(err, "Failed to get id out of assign")
+		return reconcile.Result{}, err
+	}
+
 	if deleted {
-		id, err := types.MakeID(assignMetadata)
-		if err != nil {
-			log.Error(err, "Failed to get id out of metadata")
-			return ctrl.Result{}, nil
-		}
+		tracker.CancelExpect(assignMetadata)
 
-		if err := r.system.Remove(id); err != nil {
+		if err := r.system.Remove(mID); err != nil {
 			log.Error(err, "Remove failed", "resource", request.NamespacedName)
+			return reconcile.Result{}, err
+		}
+
+		sName, err := statusv1beta1.KeyForMutatorID(util.GetPodName(), mID)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		status := &statusv1beta1.MutatorPodStatus{}
+		status.SetName(sName)
+		status.SetNamespace(util.GetNamespace())
+		if err := r.Delete(ctx, status); err != nil {
+			if !errors.IsNotFound(err) {
+				return reconcile.Result{}, err
+			}
 		}
-		return ctrl.Result{}, nil
+
+		return reconcile.Result{}, nil
+	}
+
+	status, err := r.getOrCreatePodStatus(mID)
+	if err != nil {
+		log.Info("could not get/create pod status object", "error", err)
+		return reconcile.Result{}, err
 	}
+	status.Status.MutatorUID = assignMetadata.GetUID()
+	status.Status.ObservedGeneration = assignMetadata.GetGeneration()
+	status.Status.Errors = nil
 
-	mutator, err := mutation.MutatorForAssignMetadata(assignMetadata)
+	mutator, err := mutators.MutatorForAssignMetadata(assignMetadata)
 	if err != nil {
 		log.Error(err, "Creating mutator for resource failed", "resource", request.NamespacedName)
-		tracker.CancelExpect(assignMetadata)
-		return ctrl.Result{}, err
-	}
-	if !deleted {
-		if err := r.system.Upsert(mutator); err != nil {
-			log.Error(err, "Insert failed", "resource", request.NamespacedName)
-			tracker.TryCancelExpect(assignMetadata)
-		} else {
-			tracker.Observe(assignMetadata)
+		tracker.TryCancelExpect(assignMetadata)
+		status.Status.Errors = append(status.Status.Errors, statusv1beta1.Error{Message: err.Error()})
+		if err2 := r.Update(ctx, status); err != nil {
```

Comment:
r.writer.Update?

Reply:
This reconcile struct doesn't have a reader/writer.

That was a change Oren made to certain controllers a while back "because the manager's default client bypasses the cache for unstructured resources" (per a comment he left).

This is not a concern for Assign/AssignMetadata as we are working with the typed client (not the unstructured client).

Default kubebuilder-generated code follows the present model of embedding the client in the reconciler struct. Since there is no functional difference, I'm ambivalent as to which is better. Do you prefer aligning with Kubebuilder or our other controllers?

Reply:
> This is not a concern for Assign/AssignMetadata as we are working with the typed client (not the unstructured client).

Agreed. Since this is not unstructured client, there's no functional difference. Let's keep it as is then.